code/parachain/frame/cosmwasm/src/lib.rs

```diff
@@ -674,21 +749,70 @@ pub mod pallet {
 			}
 		}
 
-		/// Handy wrapper to set the contract info
+		/// Set the contract info and update the state accordingly.
+		///
+		/// This function will update the state if the `code_id` is changing:
+		/// 1. Refcount of the new `code_id` is incremented.
+		/// 2. Refcount of the old `code_id` is decremented.
+		/// 3. Delete every entry related to old `code_id` if
+		///    the refcount is 0. And unreserve the bonded funds.
 		pub(crate) fn do_set_contract_meta(
 			contract: &AccountIdOf<T>,
 			code_id: CosmwasmCodeId,
 			admin: Option<AccountIdOf<T>>,
 			label: String,
 		) -> Result<(), Error<T>> {
 			let mut info = Self::contract_info(contract)?;
+
+			if info.code_id != code_id {
+				// Increase the refcount of `new_code_id`.
+				CodeIdToInfo::<T>::try_mutate_exists(code_id, |entry| -> Result<(), Error<T>> {
+					let code_info = entry.as_mut().ok_or(Error::<T>::CodeNotFound)?;
+					code_info.refcount =
+						code_info.refcount.checked_add(1).ok_or(Error::<T>::RefcountOverflow)?;
+					Ok(())
+				})?;
+
+				// Modify the exising `code_id`'s states and unreserve the bonded funds.
```

Comment:
```suggestion
				// Modify the existing `code_id`'s states and unreserve the bonded funds.
```
Nit spelling; failing spellcheck CI